libs/vkd3d/command.c

```diff
@@ -8345,121 +8345,559 @@ static void STDMETHODCALLTYPE d3d12_command_list_CopyTiles(d3d12_command_list_if
-    vk_image_barriers[0].oldLayout = writes_full_subresource ? VK_IMAGE_LAYOUT_UNDEFINED : dst_resource->common_layout;
-    vk_image_barriers[0].newLayout = dst_layout;
-    vk_image_barriers[0].image = dst_resource->res.vk_image;
-    vk_image_barriers[0].subresourceRange = vk_subresource_range_from_layers(&resolve->dstSubresource);
+static void d3d12_command_list_execute_resolve(struct d3d12_command_list *list,
+        struct d3d12_resource *dst_resource, struct d3d12_resource *src_resource,
+        uint32_t region_count, const VkImageResolve2 *regions, DXGI_FORMAT format,
+        D3D12_RESOLVE_MODE mode, enum vkd3d_resolve_path path)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &list->device->vk_procs;
+    struct vkd3d_resolve_image_pipeline_key resolve_pipeline_key;
+    struct vkd3d_texture_view_desc dst_view_desc, src_view_desc;
+    struct vkd3d_resolve_image_info resolve_pipeline_info;
+    struct vkd3d_resolve_image_args resolve_args;
+    VkRenderingAttachmentInfo attachment_info;
+    VkWriteDescriptorSet vk_descriptor_write;
+    struct vkd3d_view *dst_view, *src_view;
+    const struct vkd3d_format *vk_format;
+    VkDescriptorImageInfo vk_image_info;
+    VkResolveImageInfo2 resolve_info;
+    VkRenderingInfo rendering_info;
+    VkViewport viewport;
+    unsigned int i, j;
 
-    vk_image_barriers[1].dstAccessMask = VK_ACCESS_2_TRANSFER_READ_BIT;
-    vk_image_barriers[1].oldLayout = src_resource->common_layout;
-    vk_image_barriers[1].newLayout = src_layout;
-    vk_image_barriers[1].image = src_resource->res.vk_image;
-    vk_image_barriers[1].subresourceRange = vk_subresource_range_from_layers(&resolve->srcSubresource);
+    if (path == VKD3D_RESOLVE_PATH_DIRECT)
+    {
+        memset(&resolve_info, 0, sizeof(resolve_info));
+        resolve_info.sType = VK_STRUCTURE_TYPE_RESOLVE_IMAGE_INFO_2;
+        resolve_info.srcImage = src_resource->res.vk_image;
+        resolve_info.srcImageLayout = d3d12_resource_pick_layout(src_resource, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL);
+        resolve_info.dstImage = dst_resource->res.vk_image;
+        resolve_info.dstImageLayout = d3d12_resource_pick_layout(dst_resource, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL);
+        resolve_info.regionCount = region_count;
+        resolve_info.pRegions = regions;
+
+        VK_CALL(vkCmdResolveImage2(list->cmd.vk_command_buffer, &resolve_info));
+    }
+    else
+    {
+        d3d12_command_list_invalidate_current_pipeline(list, true);
+        d3d12_command_list_invalidate_root_parameters(list, &list->graphics_bindings, true, &list->compute_bindings);
+        d3d12_command_list_update_descriptor_buffers(list);
+
+        vk_format = d3d12_command_list_get_resolve_format(list, dst_resource, src_resource, format);
+
+        memset(&attachment_info, 0, sizeof(attachment_info));
+        attachment_info.sType = VK_STRUCTURE_TYPE_RENDERING_ATTACHMENT_INFO;
+        attachment_info.storeOp = VK_ATTACHMENT_STORE_OP_STORE;
```

Comment:
Can we use ATTACHMENT_STORE_OP_NONE here?

Reply:
Does `NONE` preserve the existing contents of the render target?

Reply:
Yes, and it's explicitly designed to avoid dummy writes.